File: Práctica4/java/src/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Clase que representa una asociación de estaciones espaciales
 */
public class SpaceCity extends SpaceStation{
    /**
     * Base de la ciudad espacial
     */
    private SpaceStation base;
    
    /**
     * Array con el resto de estaciones que conforman la ciudad
     */
    private ArrayList<SpaceStation> collaborators;
    
    /**
     * Constructor
     * @param _base estación que será la base de la ciudad
     * @param _rest array de estaciones que conforman la ciudad
     */
    public SpaceCity(SpaceStation _base, ArrayList<SpaceStation> _rest) {
        super(_base);
        base = _base;
        collaborators = new ArrayList<>(_rest);        
    }
    
    /**
     * Get de collaborators
     * @return collaborators
     */
    public ArrayList<SpaceStation> getCollaborators(){
        return collaborators;
    }
    
    /**
     * Todas las estaciones disparan juntas
     * @return potencia de disparo
     */
    @Override
    public float fire(){
        Iterator<SpaceStation> it = collaborators.iterator();
        float factor = base.fire(); // super.fire()
        
        while(it.hasNext()){
            factor+=it.next().fire();
        }
        
        return factor;
    }
    
    /**
     * Todas las estaciones usan sus escudos en conjunto
     * @return potencia del escudo
     */
    @Override
    public float protection(){
        Iterator<SpaceStation> it = collaborators.iterator();
        float factor = base.protection();   // super.protection()
        
        while(it.hasNext()){
            factor+=it.next().protection();
        }
        
        return factor;
    }
    
    /**
     * Recibe y procesa un loot
     * @param loot loot a procesar
     * @return transformacion
     */
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public SpaceCityToUI getUIversion() {
        return new SpaceCityToUI(this);
    }
    
}

File: Práctica4/lib/SpaceCity.rb
#encoding:utf-8

require_relative 'SpaceStation'
require_relative 'SpaceCityToUI'

module Deepspace
  # CLase que representa una asociación de estaciones espaciales
  class SpaceCity < SpaceStation
    # Constrcutor
    # @param _base [SpaceStation] base de la ciudad espacial
    # @param _rest [Array<SpaceStation>] resto de estaciones aliadas
    def initialize(_base, _rest)
      copy(_base)

      # @!attribute [SpaceStation] base de la ciudad espacial
      @base = _base

      # @!attribute [Array<SpaceStation>] resto de estaciones aliadas
      @collaborators = _rest
    end

    # Getters
    attr_reader :collaborators

    # Hace que todas las estaciones aliadas disparen a la vez
    # @return [Float] potencia de disparo acumulada
    # --Overriden
    def fire
      power = @base.fire
      @collaborators.each do |station|
        power += station.fire
      end
      power
    end

    # Usa el escudo de todas las estaciones aliadas a la vez
    # @return [Float] potencia de escudo acumulada
    # --Overriden
    def protection
      energy = @base.protection
      @collaborators.each do |station|
        energy += station.protection
      end
      energy
    end

    # Recibe un loot
    # @param [Loot] loot a recibir
    # --Overriden
    def setLoot(loot)
      super
      return Transformation::NOTRANSFORM
    end

    # String representation of the object
    # @return [String] string representation
    # --Overriden
    def to_s
      getUIversion().to_s
    end

    # To UI
    def getUIversion
      return SpaceCityToUI.new(self)
    end

  end
end
